Documents/UpgradeExtensibility.md

```diff
@@ -0,0 +1,157 @@
+VFSForGit Upgrade Extensibility Design Document
+===============================
+
+Scope
+-----
+
+Enable groups to extend the Upgrade verb to install / update extra
+components besides just VFSForGit and Git.
+
+Motivation
+----------
+
+VFSForGit can exist as a component of a larger engineering
+system. Some environments could include components that are coupled
+with specific versions of VFSForGit and should be updated with with
+VFSForGit. There could be other interactions specific to the
+environment that need to be accounted for when updating VFSForGit. As
+these requirements are specific to the individual environments
+VFSForGit is being deployed into, we need a way for groups to include
+extra components and logic in the VFSForGit upgrade.
+
+Requirements
+------------
+
+* Groups can include extra components / logic in VFSForGit Upgrade
+
+* Feed needs to support advertising current available version
+
+Non-Goals
+---------
+
+* Initial configuration of the alternate feed URL
+
+* Controlling the entire package of components
+
+Design
+------
+
+VFSForGit will support an alternative "Feed" from which updates can be
+retrieved. When an alternative feed is configured, VFSForGit will
+query this feed for updates instead of the normal feed. This
+alternative feed can contain customized components in addition to the
+VFSForGit and Git installers.
+
+### Feed
+
+The alternative feed will be a NuGet feed. The feed URL is specified
+in the gvfs configuration file (gvfs.config) with the following entry:
+
+    upgrade.feedurl
+
+This is similar to how upgrade rings are specified. The upgrade.feed
+will take precedence over the upgrade ring configuration. When this
+value is configured, VFSForGit will query and download updates from
+this feed. It must be a NuGet feed.

```

Comment:
We also need to document how VFSForGit would authenticate with the NuGet feed. And how this would affect GVFS.Service's ability to do silent upgrade checks in the background.